### ISSUE TITLE
Access Log Date Formatting

### ISSUE THREAD

**Issue by derekargueta** [envoyproxy/envoy#2709]
*Title*: Access Log Date Formatting

*Description*:
Any thought on expanding the Access Log format template to allow custom date formatting? I'm thinking something like
```
%START_TIME%  // default format
%START_TIME(%Y/%m/%d %H:%M)% // Uses the specified format
```

Being able write the unix epoch instead/additionally may also be helpful.

**Comment by htuch** [comment]
This seems reasonable. Unix epoch is easy to add, I guess `strftime` format would make sense for custom formatting. Do you plan working on this or should I leave this "help wanted"?

**Comment by dio** [comment]
Probably this https://github.com/envoyproxy/envoy/issues/2692 is related? We might want to support something similar with nginx `log_format`? http://nginx.org/en/docs/http/ngx_http_log_module.html#log_format

**Comment by htuch** [comment]
@dio I don't think so, unless you're suggesting we should have one master issue tracking all the places we don't have Nginx log format, including places where we have headers that should reflect some of these variables such as https://github.com/envoyproxy/envoy/issues/2699. Probably someone should do a design doc or list of vars that cover the union of what Nginx and Envoy supports and describes which ones are implemented, which ones should be implemented and whether we offer in request headers or access log.

**Comment by derekargueta** [comment]
@htuch you can assign this to me - are we wanting a comprehensive design doc before starting?

**Comment by derekargueta** [comment]
comprehensive design doc wrt enumerating the currently supported variables vs nginx and other proxies. My main task in this issue is to get more flexible `START_TIME` formatting but happy to help in other regards as well.

**Comment by htuch** [comment]
If you just want time format, let's go with your current proposal, seems reasonable. How do you plan on dealing with escaping considerations for `%`?

**Comment by derekargueta** [comment]
Cleared out the next few days to do this.

@htuch a couple routes we can go:
1. I can modify the parsing code to be aware of when it's inside `(...)` to interpret `%` differently as the code is currently a simple `find` call to determine the start/end of a command - the downside of this IMO is that the syntax is a bit more ambiguous as `%` is now contextual. `%START_TIME(%Y/%m/%d %H:%M)%`. The upside is that it's still a super simple `strftime` call.
2. We could use a different symbol that doesn't collide with `%` and do a string replace post-parsing so that we can still plug the format directly into `strftime`. However I can't think of a good symbol without being too different. `:` would be an nice escape symbol but that symbol is commonly used in time formats. `%START_TIME(?Y/?m/?d ?H:?M)%` as an example (I wouldn't prefer `?`)
3. We could forego `%` altogether and use date formatting more similar to what [Java](https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html) and [C#](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings) use. `%START_TIME(MM/dd/yy H:mm:ss zzz)%`. Nice and clean syntax but puts a bit more code into converting to be `strftime`-friendly

My preference, in order, would be 1, 3, 2

Lastly, we also need to be able to get the `START_TIME` as an epoch, however epoch is not part of the `strftime` syntax. Any thoughts on how that would prefer to be implemented? I think it should still be in the `START_TIME` family, perhaps a special `%START_TIME(EPOCH)%` case?

**Comment by htuch** [comment]
As long as we stick with something that is basically `strftime` I think it's good, even if it's modulo some simple search-and-replace. (1) seems worth trying, I think the first opening parens `(` should make this not too hard.
